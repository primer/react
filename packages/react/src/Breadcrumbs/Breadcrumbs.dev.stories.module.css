.EditableTextInput {
  min-width: 120px;
  max-width: 180px;
  font-size: inherit;
  border: var(--borderWidth-thin) dashed var(--borderColor-muted);
}

.EditableTextInput:focus {
  border: var(--borderWidth-thin) solid var(--borderColor-accent-emphasis);
}
