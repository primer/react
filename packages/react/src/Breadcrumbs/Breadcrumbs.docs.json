{
  "id": "breadcrumbs",
  "name": "Breadcrumbs",
  "status": "alpha",
  "a11yReviewed": false,
  "stories": [],
  "importPath": "@primer/react",
  "props": [
    {
      "name": "children",
      "type": "Breadcrumbs.Item[]",
      "defaultValue": "",
      "description": ""
    },
    {
      "name": "sx",
      "type": "SystemStyleObject"
    }
  ],
  "subcomponents": [
    {
      "name": "Breadcrumbs.Item",
      "props": [
        {
          "name": "selected",
          "type": "boolean",
          "defaultValue": "false",
          "description": ""
        },
        {
          "name": "ref",
          "type": "React.RefObject<HTMLAnchorElement>"
        },
        {
          "name": "as",
          "type": "React.ElementType",
          "defaultValue": "\"a\""
        },
        {
          "name": "sx",
          "type": "SystemStyleObject"
        }
      ],
      "passthrough": {
        "element": "a",
        "url": "https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a#Attributes"
      }
    }
  ]
}