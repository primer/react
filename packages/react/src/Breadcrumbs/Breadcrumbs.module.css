.BreadcrumbsBase {
  display: flex;
  justify-content: space-between;
}

.BreadcrumbsList {
  padding-left: 0;
  margin-top: 0;
  margin-bottom: 0;
}

.ItemWrapper {
  display: inline-block;
  font-size: var(--text-body-size-medium);
  white-space: nowrap;
  list-style: none;

  &::after {
    display: inline-block;
    height: 0.8em;
    /* stylelint-disable-next-line primer/spacing */
    margin: 0 0.5em;
    font-size: var(--text-body-size-medium);
    content: '';
    /* stylelint-disable-next-line primer/borders, primer/colors */
    border-right: 0.1em solid var(--fgColor-muted);
    transform: rotate(15deg) translateY(0.0625em);
  }

  &:first-child {
    margin-left: 0;
  }

  &:last-child {
    &::after {
      content: none;
    }
  }
}

.Item {
  display: inline-block;
  font-size: var(--text-body-size-medium);
  color: var(--fgColor-link);
  text-decoration: none;

  &:hover,
  &:focus {
    text-decoration: underline;
  }
}

.ItemSelected {
  color: var(--fgColor-default);
  pointer-events: none;

  &:focus {
    text-decoration: none;
  }
}
