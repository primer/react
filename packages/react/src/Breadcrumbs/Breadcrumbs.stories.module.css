.EditableInput {
  min-width: '120px';
  max-width: '180px';
  font-size: 'inherit';
  border: 1px dashed var(--borderColor-muted);

  &:focus {
    border: 1px solid var(--borderColor-accent-emphasis);
  }
}