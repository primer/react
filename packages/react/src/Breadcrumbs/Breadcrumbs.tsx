import clsx from 'clsx'
import type {To} from 'history'
import React from 'react'
import styled from 'styled-components'
import Box from '../Box'
import {get} from '../constants'
import type {SxProp} from '../sx'
import sx from '../sx'
import type {ComponentProps} from '../utils/types'

const SELECTED_CLASS = 'selected'

const Wrapper = styled.li`
  display: inline-block;
  white-space: nowrap;
  list-style: none;
  &::after {
    font-size: ${get('fontSizes.1')};
    content: '';
    display: inline-block;
    height: 0.8em;
    margin: 0 0.5em;
    border-right: 0.1em solid;
    border-color: ${get('colors.fg.muted')};
    transform: rotate(15deg) translateY(0.0625em);
  }
  &:first-child {
    margin-left: 0;
  }
  &:last-child {
    &::after {
      content: none;
    }
  }
`

const BreadcrumbsBase = styled.nav<SxProp>`
  display: flex;
  justify-content: space-between;
  ${sx};
`

export type BreadcrumbsProps = React.PropsWithChildren<
  {
    className?: string
  } & SxProp
>

function Breadcrumbs({className, children, sx: sxProp}: React.PropsWithChildren<BreadcrumbsProps>) {
  const wrappedChildren = React.Children.map(children, child => <Wrapper>{child}</Wrapper>)
  return (
    <BreadcrumbsBase className={className} aria-label="Breadcrumbs" sx={sxProp}>
      <Box as="ol" my={0} pl={0}>
        {wrappedChildren}
      </Box>
    </BreadcrumbsBase>
  )
}

type StyledBreadcrumbsItemProps = {
  to?: To
  selected?: boolean
} & SxProp

const BreadcrumbsItem = styled.a.attrs<StyledBreadcrumbsItemProps>(props => ({
  className: clsx(props.selected && SELECTED_CLASS, props.className),
  'aria-current': props.selected ? 'page' : null,
}))<StyledBreadcrumbsItemProps>`
  color: ${get('colors.accent.fg')};
  display: inline-block;
  font-size: ${get('fontSizes.1')};
  text-decoration: none;
  &:hover,
  &:focus {
    text-decoration: underline;
  }
  &.selected {
    color: ${get('colors.fg.default')};
    pointer-events: none;
  }
  &.selected:focus {
    text-decoration: none;
  }
  ${sx};
`

Breadcrumbs.displayName = 'Breadcrumbs'

BreadcrumbsItem.displayName = 'Breadcrumbs.Item'

export type BreadcrumbsItemProps = ComponentProps<typeof BreadcrumbsItem>
export default Object.assign(Breadcrumbs, {Item: BreadcrumbsItem})

/**
 * @deprecated Use the `Breadcrumbs` component instead (i.e. `<Breadcrumb>` â†’ `<Breadcrumbs>`)
 */
export const Breadcrumb = Object.assign(Breadcrumbs, {Item: BreadcrumbsItem})

/**
 * @deprecated Use the `BreadcrumbsProps` type instead
 */
export type BreadcrumbProps = ComponentProps<typeof BreadcrumbsBase>

/**
 * @deprecated Use the `BreadcrumbsItemProps` type instead
 */
export type BreadcrumbItemProps = ComponentProps<typeof BreadcrumbsItem>
