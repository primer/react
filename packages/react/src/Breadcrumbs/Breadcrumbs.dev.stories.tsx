import Breadcrumbs from '.'

export default {
  title: 'Components/Breadcrumbs/Dev',
}

export const DynamicChildren = () => {
  const [items, setItems] = useState([
    {id: 1, href: '#', name: 'Home'},
    {id: 2, href: '#', name: 'Docs'},
    {id: 3, href: '#', name: 'Components'},
  ])

  const addItem = () => {
    const newId = Math.max(...items.map(item => item.id)) + 1
    const names = ['Advanced', 'Examples', 'Guides', 'API', 'Tutorials', 'Reference']
    const randomName = names[Math.floor(Math.random() * names.length)]
    setItems([...items, {id: newId, href: '#', name: `${randomName}-${newId}`}])
  }

  const removeItem = () => {
    if (items.length > 1) {
      setItems(items.slice(0, -1))
    }
  }

  const addMultipleItems = () => {
    const newItems = [
      {id: Date.now() + 1, href: '#', name: 'Category'},
      {id: Date.now() + 2, href: '#', name: 'Subcategory'},
      {id: Date.now() + 3, href: '#', name: 'Item'},
      {id: Date.now() + 4, href: '#', name: 'Details'},
      {id: Date.now() + 5, href: '#', name: 'Specifications'},
    ]
    setItems([...items, ...newItems])
  }

  const reset = () => {
    setItems([
      {id: 1, href: '#', name: 'Home'},
      {id: 2, href: '#', name: 'Docs'},
      {id: 3, href: '#', name: 'Components'},
    ])
  }

  return (
    <div style={{display: 'flex', flexDirection: 'column', gap: '16px'}}>
      <div style={{display: 'flex', gap: '8px', marginBottom: '16px'}}>
        <button type="button" onClick={addItem} style={{padding: '4px 8px'}}>
          Add Item
        </button>
        <button type="button" onClick={removeItem} style={{padding: '4px 8px'}}>
          Remove Item
        </button>
        <button type="button" onClick={addMultipleItems} style={{padding: '4px 8px'}}>
          Add Many Items
        </button>
        <button type="button" onClick={reset} style={{padding: '4px 8px'}}>
          Reset
        </button>
      </div>

      <div>
        <h4 id="dynamic-breadcrumbs-heading" style={{margin: '0 0 8px 0'}}>
          Dynamic breadcrumbs
        </h4>
        <Breadcrumbs overflow="menu-with-root">
          {items.map((item, index) => (
            <Breadcrumbs.Item key={item.id} href={item.href} selected={index === items.length - 1}>
              {item.name}
            </Breadcrumbs.Item>
          ))}
        </Breadcrumbs>
      </div>

      <div style={{marginTop: '16px', fontSize: '12px'}}>
        Current items: {items.length} | Try adding/removing items to see how overflow behavior changes
      </div>
    </div>
  )
}

export const OverflowMenuNarrowContainer = () => (
  <div style={{width: '350px', border: '1px solid #ccc', padding: '8px'}}>
    <Breadcrumbs overflow="menu">
      <Breadcrumbs.Item href="#">Home</Breadcrumbs.Item>
      <Breadcrumbs.Item href="#">Products</Breadcrumbs.Item>
      <Breadcrumbs.Item href="#">Category</Breadcrumbs.Item>
      <Breadcrumbs.Item href="#">Subcategory</Breadcrumbs.Item>
      <Breadcrumbs.Item href="#" selected>
        Current Page
      </Breadcrumbs.Item>
    </Breadcrumbs>
  </div>
)

// Wrapper components to test that BreadcrumbsItem works when wrapped
const StyledWrapper = ({children}: {children: React.ReactNode}) => (
  <span style={{padding: '2px', border: '1px dotted #999'}}>{children}</span>
)

const ConditionalWrapper = ({children, condition}: {children: React.ReactNode; condition: boolean}) => {
  return condition ? <strong>{children}</strong> : <>{children}</>
}

const DataAttributeWrapper = ({children}: {children: React.ReactNode}) => (
  <span data-testid="wrapper" className="custom-wrapper">
    {children}
  </span>
)

export const WrappedBreadcrumbItemsWithOverflow = () => (
  <Breadcrumbs overflow="menu">
    <StyledWrapper>
      <Breadcrumbs.Item href="#">Wrapped Home</Breadcrumbs.Item>
    </StyledWrapper>
    <ConditionalWrapper condition={false}>
      <Breadcrumbs.Item href="#">Products</Breadcrumbs.Item>
    </ConditionalWrapper>
    <DataAttributeWrapper>
      <Breadcrumbs.Item href="#">Category</Breadcrumbs.Item>
    </DataAttributeWrapper>
    <StyledWrapper>
      <Breadcrumbs.Item href="#">Subcategory</Breadcrumbs.Item>
    </StyledWrapper>
    <ConditionalWrapper condition={true}>
      <Breadcrumbs.Item href="#">Item</Breadcrumbs.Item>
    </ConditionalWrapper>
    <DataAttributeWrapper>
      <Breadcrumbs.Item href="#">Details</Breadcrumbs.Item>
    </DataAttributeWrapper>
    <Breadcrumbs.Item href="#" selected>
      Current Page
    </Breadcrumbs.Item>
  </Breadcrumbs>
)

export const WithEditableNameInput = () => (
  <Breadcrumbs>
    <Breadcrumbs.Item href="#">Home</Breadcrumbs.Item>
    <Breadcrumbs.Item href="#">Documents</Breadcrumbs.Item>
    <Breadcrumbs.Item href="#">Project Alpha</Breadcrumbs.Item>
    <Breadcrumbs.Item>
      <TextInput
        defaultValue="Untitled Document"
        size="small"
        sx={{
          minWidth: '120px',
          maxWidth: '180px',
          fontSize: 'inherit',
          border: '1px dashed var(--borderColor-muted)',
          '&:focus': {
            border: '1px solid var(--borderColor-accent-emphasis)',
          },
        }}
        aria-label="Edit document name"
      />
    </Breadcrumbs.Item>
  </Breadcrumbs>
)
