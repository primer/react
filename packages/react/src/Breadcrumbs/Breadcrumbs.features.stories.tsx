import type {Meta} from '@storybook/react-vite'
import type React from 'react'
import type {ComponentProps} from '../utils/types'
import Breadcrumbs from './Breadcrumbs'

export default {
  title: 'Components/Breadcrumbs/Features',
  component: Breadcrumbs,
} as Meta<ComponentProps<typeof Breadcrumbs>>

export const OverflowWrap = () => (
  <Breadcrumbs overflow="wrap">
    <Breadcrumbs.Item href="#">Home</Breadcrumbs.Item>
    <Breadcrumbs.Item href="#">Products</Breadcrumbs.Item>
    <Breadcrumbs.Item href="#">Category</Breadcrumbs.Item>
    <Breadcrumbs.Item href="#">Subcategory</Breadcrumbs.Item>
    <Breadcrumbs.Item href="#">Item</Breadcrumbs.Item>
    <Breadcrumbs.Item href="#">Details</Breadcrumbs.Item>
    <Breadcrumbs.Item href="#" selected>
      Current Page
    </Breadcrumbs.Item>
  </Breadcrumbs>
)

export const OverflowMenu = () => (
  <Breadcrumbs overflow="menu">
    <Breadcrumbs.Item href="#">Home</Breadcrumbs.Item>
    <Breadcrumbs.Item href="#">Products</Breadcrumbs.Item>
    <Breadcrumbs.Item href="#">Category</Breadcrumbs.Item>
    <Breadcrumbs.Item href="#">SubcategorySubcategorySubcategorySubcategory</Breadcrumbs.Item>
    <Breadcrumbs.Item href="#">ItemItemItemItemItemItemItem</Breadcrumbs.Item>
    <Breadcrumbs.Item href="#">DetailsDetailsDetailsDetailsDetailsDetailsDetails</Breadcrumbs.Item>
    <Breadcrumbs.Item href="#" selected>
      Current Page
    </Breadcrumbs.Item>
  </Breadcrumbs>
)

export const OverflowMenuShowRoot = () => (
  <Breadcrumbs overflow="menu" hideRoot={false}>
    <Breadcrumbs.Item href="#">Home</Breadcrumbs.Item>
    <Breadcrumbs.Item href="#">Products</Breadcrumbs.Item>
    <Breadcrumbs.Item href="#">Category</Breadcrumbs.Item>
    <Breadcrumbs.Item href="#">Subcategory</Breadcrumbs.Item>
    <Breadcrumbs.Item href="#">Item</Breadcrumbs.Item>
    <Breadcrumbs.Item href="#">Details</Breadcrumbs.Item>
    <Breadcrumbs.Item href="#" selected>
      Current Page
    </Breadcrumbs.Item>
  </Breadcrumbs>
)

export const OverflowMenuNarrowContainer = () => (
  <div style={{width: '350px', border: '1px solid #ccc', padding: '8px'}}>
    <Breadcrumbs overflow="menu">
      <Breadcrumbs.Item href="#">Home</Breadcrumbs.Item>
      <Breadcrumbs.Item href="#">Products</Breadcrumbs.Item>
      <Breadcrumbs.Item href="#">Category</Breadcrumbs.Item>
      <Breadcrumbs.Item href="#">Subcategory</Breadcrumbs.Item>
      <Breadcrumbs.Item href="#" selected>
        Current Page
      </Breadcrumbs.Item>
    </Breadcrumbs>
  </div>
)

// Wrapper components to test that BreadcrumbsItem works when wrapped
const StyledWrapper = ({children}: {children: React.ReactNode}) => (
  <span style={{padding: '2px', border: '1px dotted #999'}}>{children}</span>
)

const ConditionalWrapper = ({children, condition}: {children: React.ReactNode; condition: boolean}) => {
  return condition ? <strong>{children}</strong> : <>{children}</>
}

const DataAttributeWrapper = ({children}: {children: React.ReactNode}) => (
  <span data-testid="wrapper" className="custom-wrapper">
    {children}
  </span>
)

export const WrappedBreadcrumbItemsWithOverflow = () => (
  <Breadcrumbs overflow="menu">
    <StyledWrapper>
      <Breadcrumbs.Item href="#">Wrapped Home</Breadcrumbs.Item>
    </StyledWrapper>
    <ConditionalWrapper condition={false}>
      <Breadcrumbs.Item href="#">Products</Breadcrumbs.Item>
    </ConditionalWrapper>
    <DataAttributeWrapper>
      <Breadcrumbs.Item href="#">Category</Breadcrumbs.Item>
    </DataAttributeWrapper>
    <StyledWrapper>
      <Breadcrumbs.Item href="#">Subcategory</Breadcrumbs.Item>
    </StyledWrapper>
    <ConditionalWrapper condition={true}>
      <Breadcrumbs.Item href="#">Item</Breadcrumbs.Item>
    </ConditionalWrapper>
    <DataAttributeWrapper>
      <Breadcrumbs.Item href="#">Details</Breadcrumbs.Item>
    </DataAttributeWrapper>
    <Breadcrumbs.Item href="#" selected>
      Current Page
    </Breadcrumbs.Item>
  </Breadcrumbs>
)
