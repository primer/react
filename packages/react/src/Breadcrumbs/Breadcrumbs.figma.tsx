import React from 'react'
import Breadcrumbs from './Breadcrumbs'
import figma from '@figma/code-connect'

/**
 * -- This file was auto-generated by `figma connect create` --
 * `props` includes a mapping from Figma properties and variants to
 * suggested values. You should update this to match the props of your
 * code component, and update the `example` function to return the
 * code example you'd like to see in Figma
 */

figma.connect(
  Breadcrumbs,
  'https://www.figma.com/design/GCvY3Qv8czRgZgvl1dG6lp/Primer-Web?m=auto&node-id=20135-70690&t=39jWyeflbJqVh77d-1',
  {
    props: {
      children: figma.children([
        '1st parent',
        '2nd parent',
        '3rd parent',
        '4th parent',
        '5th parent',
        '6th parent',
        '7th parent',
        '8th parent',
        '9th parent',
        'Current page',
      ]),
    },
    example: ({children}) => <Breadcrumbs>{children}</Breadcrumbs>,
  },
)

/**
 * BreadcrumbsItem
 */
figma.connect(Breadcrumbs.Item, 'https://www.figma.com/design/GCvY3Qv8czRgZgvl1dG6lp/Primer-Web?node-id=34258-144883', {
  props: {
    selected: figma.boolean('selected'),
    label: figma.textContent('label'),
  },
  example: ({selected, label}) => (
    <Breadcrumbs.Item href="#" selected={selected}>
      {label}
    </Breadcrumbs.Item>
  ),
})
