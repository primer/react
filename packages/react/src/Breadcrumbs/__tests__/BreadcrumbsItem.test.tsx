import {render as HTMLRender} from '@testing-library/react'
import axe from 'axe-core'
import React from 'react'
import Breadcrumbs from '..'
import {behavesAsComponent, render} from '../../utils/testing'

describe('Breadcrumbs.Item', () => {
  behavesAsComponent({Component: Breadcrumbs.Item})

  it('renders an <a> by default', () => {
    expect(render(<Breadcrumbs.Item />).type).toEqual('a')
  })

  it('should have no axe violations', async () => {
    const {container} = HTMLRender(<Breadcrumbs.Item />)
    const results = await axe.run(container)
    expect(results).toHaveNoViolations()
  })

  it('respects the "selected" prop', () => {
    expect(render(<Breadcrumbs.Item selected />)).toMatchSnapshot()
  })
})
