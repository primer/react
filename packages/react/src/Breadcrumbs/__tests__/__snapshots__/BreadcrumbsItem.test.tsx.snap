// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumbs.Item respects the "selected" prop 1`] = `
.c0 {
  color: var(--fgColor-accent,var(--color-accent-fg,#0969da));
  display: inline-block;
  font-size: 14px;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0:hover,
.c0:focus {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c0.selected {
  color: var(--fgColor-default,var(--color-fg-default,#1F2328));
  pointer-events: none;
}

.c0.selected:focus {
  -webkit-text-decoration: none;
  text-decoration: none;
}

<a
  aria-current="page"
  className="c0 selected"
  selected={true}
/>
`;
